### ISSUE TITLE
ERROR: pgaudit stack is not empty

### ISSUE THREAD

**Issue by Giak79** [pgaudit/pgaudit#64]
HI,

I have a problem: with PGAUDIT  VS Oracle Data Integrator.

ODI generates the following error when reading the data contained in the PostgreSQL table:

(ODI-1227: Task SrcSet0 (loading) made on the STOREPICKING PostgreSQL source connection.
  Caused by: org.postgresql.util.PSQLException: ERROR: pgaudit stack is not empty)

In PGAUDIT log file there is the same error without any specification.

With PGAUDIT in disable mode ( removing also from shared_preload_libraries  ) work fine .

PostgreSQL driver: postgresql-42.2.5.jre6.jar
Oracle Data Integrator: 11.1.1.1.6.0
DB PostgreSQL 9.6.10
PGAUDIT : V1.1.X

Thank's in advances

**Comment by dwsteele** [comment]
We don't have access to ODI to test this (and no repro is given) so it's hard to say what might be going on.  If you can isolate the exact SQL statement or statements that are causing the problem then we'll have a look.

**Comment by HesusFTW** [comment]
Hi @dwsteele this defect is very easy to reproduce, it seems to be related to setting the fetchSize (cursor size) in the postgresql jdbc driver.
Using the following java code:
PreparedStatement preparedStatement = dbConnection.prepareStatement(query, ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_UPDATABLE);
// if the bellow line is uncommented then it gives an error:
//			preparedStatement.setFetchSize(maxRowsToFetch); 

When I try to commit the DB i get:
	Caused by: org.postgresql.util.PSQLException: ERROR: pgaudit stack is not empty
		at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2440)
		at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2183)
		at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:308)
		at org.postgresql.jdbc.PgConnection.executeTransactionCommand(PgConnection.java:755)
		at org.postgresql.jdbc.PgConnection.commit(PgConnection.java:777)

I am using the latest jdbc driver:
<dependency>
			<groupId>org.postgresql</groupId>
			<artifactId>postgresql</artifactId>
			<version>42.2.5</version>
		</dependency>
The database is: 10.7
The pgAudit v1.2.X is intended to support PostgreSQL (i think i don't have a hard confirmation yet, will update when i have complete info)

**Comment by laurenz** [comment]
One of our customers is experiencing the same problem.
The error message is only thrown if `setFetchSize` is called with a non-zero value.

Could you reopen the issue and mark it as a bug, or is it better if I create a new issue?

**Comment by dwsteele** [comment]
Hi Laurenz -- I think it's fine to use this issue.

Do you know what underlying Postgres commands are being called here?

**Comment by laurenz** [comment]
When the program calls `setFetchSize` with a size > 0, the JDBC driver uses the extended query protocol with a portal (cursor) on an unnamed prepared statement in PostgreSQL.

The log file will look like this if you set `log_min_duration_statement = 0` (fetch size 20, no pgAudit running here):

    parse <unnamed>: BEGIN
    bind <unnamed>: BEGIN
    execute <unnamed>: BEGIN
    parse <unnamed>: SELECT id FROM tab_a
    bind <unnamed>/C_1: SELECT id FROM tab_a
    execute <unnamed>/C_1: SELECT id FROM tab_a
    execute fetch from <unnamed>/C_1: SELECT id FROM tab_a
    execute fetch from <unnamed>/C_1: SELECT id FROM tab_a
    execute fetch from <unnamed>/C_1: SELECT id FROM tab_a
    execute fetch from <unnamed>/C_1: SELECT id FROM tab_a
    execute fetch from <unnamed>/C_1: SELECT id FROM tab_a

This corresponds to the following protocol messages sent from the frontend to the backend:

- Parse (`P`) with an empty string for the statement name and the query
- Bind (`B`) with `C_1` for the portal, an empty statement name and 0 for all other arguments (no parameters)
- Describe (`D`) with a `P` for "portal" and the portal name `C_1`
- Execute (`E`) with the portal name `C_1` and the number of rows 20
- Sync (`S`) to signify that it waits for a result from the backend

The backend sends the data, followed by &ldquo;Portal suspend&rdquo; (`s`) and &ldquo;Ready for query&rdquo (`Z`).

The frontend keeps repeating &ldquo;Execute&rdquo; and &ldquo;Sync&rdquo; until the backend responds with &ldquo;CommandComplete&rdquo; (`C`).

I described the protocol in such detail because libpq does not support portals, but it can be reproduced easily with a simple Java program like this:

```java
public class UseCursor {
    public static void main(String[] args) throws ClassNotFoundException,
                                           java.sql.SQLException {
        Class.forName("org.postgresql.Driver");
        java.sql.Connection conn =
            java.sql.DriverManager.getConnection("jdbc:postgresql://localhost/dbname");
        conn.setAutoCommit(false);
        java.sql.PreparedStatement stmt = conn.prepareStatement("SELECT id FROM tab_a");
        stmt.setFetchSize(20);
        java.sql.ResultSet rs = stmt.executeQuery();
        while (rs.next())
            System.out.println(rs.getString(1));
        rs.close();
        stmt.close();
        conn.close();
    }
}
```

I didn't try to reproduce the problem with pgAudit myself, but perhaps you don't need JDBC, but can use e.g. a cursor in PL/pgSQL.

**Comment by laurenz** [comment]
The actual code that caused the error at the customer was:

```java
preparedStatement = dbConnection.prepareStatement(
                        query,
                        ResultSet.TYPE_FORWARD_ONLY,
                        ResultSet.CONCUR_UPDATABLE
                    );
preparedStatement.setFetchSize(10);
```

**Comment by dwsteele** [comment]
@laurenz @HesusFTW I have run Laurenz' test program on the latest releases of 9.6, 10, and 11 using the latest JDBC driver and pgaudit and I don't see any issues. 

Here are the pgaudit settings I used:
```
shared_preload_libraries = 'pgaudit'
pgaudit.log = 'all'
```
The Java program:
```
public class UseCursor {
    public static void main(String[] args) throws ClassNotFoundException, java.sql.SQLException
    {
        Class.forName("org.postgresql.Driver");
        java.sql.Connection conn =
            java.sql.DriverManager.getConnection("jdbc:postgresql://localhost/postgres");
        conn.setAutoCommit(false);
        java.sql.PreparedStatement stmt = conn.prepareStatement(
            "SELECT * FROM pg_class", ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_UPDATABLE);
        stmt.setFetchSize(20);
        java.sql.ResultSet rs = stmt.executeQuery();
        while (rs.next())
            System.out.println(rs.getString(1));
        rs.close();
        stmt.close();
        conn.close();
    }
}
```
All versions gave the same output:
```
AUDIT: SESSION,1,1,MISC,SET,,,SET extra_float_digits = 3,<not logged>
AUDIT: SESSION,2,1,MISC,SET,,,SET application_name = 'PostgreSQL JDBC Driver',<not logged>
LOG:  AUDIT: SESSION,3,1,MISC,BEGIN,,,BEGIN,<not logged>
LOG:  AUDIT: SESSION,4,1,READ,SELECT,,,SELECT * FROM pg_class,<not logged>
```
I also tried enabling some extra options on v10 which changed the logging slightly but I didn't see any issues:
```
pgaudit.log_parameter = on
pgaudit.log_relation = on
```
So it looks like I'll need a solid repro in order to make a progress on this.

**Comment by laurenz** [comment]
Thanks for looking into that.

The customer that alerted me about that has asked me not to pursue the matter any further, so I cannot provide a reproducer.

Maybe it is best to close the issue again until somebody comes who can reproduce it.

**Comment by dwsteele** [comment]
OK, thanks for replying.  I'll close this issue until somebody comes up with a reproducible test case.

**Comment by svb007** [comment]
In our case it was a function that was erroneously declared as IMMUTABLE

**Comment by dwsteele** [comment]
Interesting -- I'm not sure why that would be a problem but it's something to try.

**Comment by svb007** [comment]
I managed to trigger this on PG11 with the following:

begin;
create table pgaud_test1 as select * from generate_series(1,200) c1,generate_series('2019-01-01'::timestamp,'2019-01-31'::timestamp,'1 day'::interval) c2;
create table pgaud_test2 as select * from generate_series(1,200) c1,generate_series('2019-01-01'::timestamp,'2019-01-31'::timestamp,'1 day'::interval) c2;
CREATE OR REPLACE FUNCTION get_pgaud_ids(_ret REFCURSOR) RETURNS REFCURSOR LANGUAGE plpgsql IMMUTABLE AS 'BEGIN OPEN _ret FOR SELECT max(c1) FROM pgaud_test1 join pgaud_test2 using (c1);RETURN _ret;END;';
select get_pgaud_ids('_ret');
fetch all from _ret;

**Comment by svb007** [comment]
Obviously, removing the IMMUTABLE flag resolves the problem

**Comment by svb007** [comment]
I'm getting the error already at the point where session_preload_libraries is set to pgaudit, even before creating the extension!

**Comment by svb007** [comment]
This is probably worse than I thought, because it also crashes when the function is actually a valid immutable function returning a cursor:
begin;
CREATE OR REPLACE FUNCTION get_pgaud_ids(_ret REFCURSOR) RETURNS REFCURSOR LANGUAGE plpgsql IMMUTABLE AS 'BEGIN OPEN _ret FOR SELECT 200 ;RETURN _ret;END;';
select get_pgaud_ids('_ret');
fetch all from _ret;

**Comment by dwsteele** [comment]
We'll reopen the issue, but not sure when we'll have a chance to debug it.  It would be helpful if you could reproduce the error in the tests at: https://github.com/pgaudit/pgaudit/blob/master/sql/pgaudit.sql

**Comment by dylrich** [comment]
Ran into this issue today in Postgres 11 ([the initial stack trace I got more or less matched this related issue](https://github.com/MagicStack/asyncpg/issues/272)). Turning off pgaudit fixed the issue for now, but obviously this isn't ideal!

**Comment by svb007** [comment]
> Ran into this issue today in Postgres 11 ([the initial stack trace I got more or less matched this related issue](https://github.com/MagicStack/asyncpg/issues/272)). Turning off pgaudit fixed the issue for now, but obviously this isn't ideal!

I did narrow it down specifically to IMMUTABLE functions returning refcursor. A better workaround would be to change those functions to VOLATILE or STABLE

**Comment by dylrich** [comment]
I am not doing that myself explicitly, perhaps `asyncpg` does underneath the hood? It does only happen when I try to use a cursor in that library, and at the very end after iterating through the cursor.

**Comment by yugo-n** [comment]
@dwsteele Attached is a Java code to reproduce "ERROR :pgaudit stack is not empty" using JDBC.

pgaudit assumes that the internal stack is empty when a utility command like CREATE or DROP arrives and raises an error if this is not empty. Events in the stack are cleared when PostgreSQL drops a portal via MemoryContextCallback, and a named portal is dropped when a "close portal" message is received. Hence, this error occurs when a new utility command is issued after a named portal is created but is not closed.

In the attached code, the error will occur if "s2.close()" is commented out, but not if s2 is closed since Close message is sent.

[Test1.java.txt](https://github.com/pgaudit/pgaudit/files/4544117/Test1.java.txt)

**Comment by dwsteele** [comment]
See PR #108 for a possible solution. Please test in your environments and let us know the results.

**Comment by yugo-n** [comment]
Sorry, the previous attached test was wrong. This causes the following error if pgaudit  is not used.
```
ERROR: cannot DROP TABLE "t2" because it is being used by active queries in this session
```

So, I tested #108 using the new test attached in this post, in which I use ANALYZE instead of DROP TABLE.  Using this test, I confirmed that it works well for the SELECT statement case.  However, it raises the error when I used SHOW or EXPLAIN instead of SELECT like commented-out SQL2 in the attached test.  Maybe this is a corner case, but I think it is not enough to allow only SELECT statement in the stacked  items.


[Test2.java.txt](https://github.com/pgaudit/pgaudit/files/4816472/Test2.java.txt)

**Comment by dwsteele** [comment]
@yugo-n I have been unable to reproduce the explain/show problems in our test environment (psql expect tests), but I added show and explain to the exceptions list in #108. Can you try again and see if it works?

**Comment by yugo-n** [comment]
Thank you for fixing this. I have confirmed the latest code fixes the issue with EXPLAIN and SHOW.

**Comment by dwsteele** [comment]
This fix has now been committed (437a537345b38613f) and back-patched to all branches.

**Comment by dylrich** [comment]
Thanks @dwsteele!
